### PR TITLE
Best practices report

### PR DESCRIPTION
**Pull request by shivanshuag**
Issue #14 
### Sites

Checks if `sites/sites.php` is a symbolic link
- [x] Pass - Either `sites.php` does not exist or `sites.php` exists and is not a symbolic link.
- [x] Warn - `sites.php` exists and is a symbolic link.
### Multisite

There is just a small difference between Drupal 7 and Drupal 8 multisite. In Drupal 8, `sites.php` file must exist in `sites` directory to enable multisite feature. Therefore, I have put an additional if statement in the code to check for the file.
- [x] Pass - do not modify the sites directory
- [x] Info - run these commands inside the sites directory `cp example.sites.php sites.php` and `mkdir site1`
- [x] Warn - Either `sites.php` is present and multisite directories not present or the other way round
- [x] Fail - run with multisite enabled and option --vendor=pantheon
### Settings

No changes from Drupal 7 version
- [x] Pass - run on default installation
- [x] Warn - make `settings.php` file inside `sites/default` a symlink
- [x] Fail - when settings.php file is not present. Drush fails to bootstrap in this case, so site_audit should fail to run
### Services

Similar to Settings check
- [x] Pass - run on default installation. `services.yml` file is present inside `sites/default` directory and is not a symlink
- [x] Warn - make `services.yml` file inside `sites/default` a symlink
- [x] Fail - delete the  `services.yml` file from `sites/default`  directory
## SitesDefault
- [x] Pass - run on default installation
- [x] Warn - make `sites/default` directory a symlink
- [x] Fail - Remove `sites/default` directory
## SitesSuperfluous
- [x] Pass - run on default installation
- [x] Warn - create any file other than `sites.php` inside `sites` directory.
### Sites All/Default/Superfluous
- `/sites/all/{modules,themes,libraries,profiles}` directories from Drupal 7 has been moved to `/{modules,themes,libraries,profiles}` in Drupal 8
- `sites/all` is still supported in Drupal 8, but not needed.
- `sites/default` is same as Drupal 7
- There are new files inside sites directory - `development.services.yml` and `example.settings.local.php`.
- Recommendations - No need for SitesAll check. We can have a check for `/{modules,themes,libraries,profiles}` direcotries. Keep SitesDefault and SitesSuperfluous checks same.

### CODE REVIEW

File: Check/BestPractices/Multisite.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * Implements \SiteAudit\Check\Abstract\getResultWarn().
+   */
+  public function getResultWarn() {}
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getAction().
+   */
+  public function getAction() {
+    if ($this->score == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_FAIL) {
+      return dt('See https://www.getpantheon.com/blog/much-ado-about-drupal-multisite for details.');
```
Comment (fluxsauce) [inline Check/BestPractices/Multisite.php]: They updated - https://pantheon.io/blog/drupal-multisite-much-ado-about-drupal-multisite

File: Check/BestPractices/Multisite.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\calculateScore().
+   */
+  public function calculateScore() {
+    $drupal_root = drush_get_context('DRUSH_SELECTED_DRUPAL_ROOT');
+    if (is_file($drupal_root . '/sites/sites.php')) {
```
Comment (fluxsauce) [inline Check/BestPractices/Multisite.php]: Always do the check regardless of if `/sites/sites.php` is a file; warn if there are directories but no `sites.php`. Also, warn if `/sites/sites.php` is a symbolic link.

File: Check/BestPractices/Services.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getAction().
+   */
+  public function getAction() {
+    if ($this->score == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_WARN) {
+      return dt('Don\'t rely on symbolic links for core configuration files; copy services.yml where it should be and remove the symbolic link.');
+    }
+    if ($this->score == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_FAIL) {
+      return dt('Create services.yml file inside sites/default directory by copying default.services.yml');
```
Comment (fluxsauce) [inline Check/BestPractices/Services.php]: > Create a services.yml file inside the sites/default directory by copying the default.services.yml file. See https://www.drupal.org/documentation/install/settings-file for details.

**Comment by fluxsauce** [comment]
Code looks really good, a couple minor suggestions. Once those are implemented, I'll functionally test. Good work so far, great to see a PR so soon!

File: Check/BestPractices/Multisite.php (hunk 2 above)
Comment (shivanshuag) [inline Check/BestPractices/Multisite.php]: I think checking for sites.php to be a symbolic link is a separate check. Giving two types of recommendations in multisite check will make it confusing. What do you suggest?

Comment (fluxsauce) [inline Check/BestPractices/Multisite.php]: Agree, separate check is logical, sequence it before the multisite check. That way, you can just warn if there is a `sites.php` by using a registry stored value.

File: Check/BestPractices/Multisite.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  public function getResultPass() {
+    return dt('No multi-sites detected.');
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultWarn().
+   */
+  public function getResultWarn() {
+    if ($this->registry['multisite_enabled']) {
+      return dt('sites/sites.php is present but no multisite directory present.');
```
Comment (fluxsauce) [inline Check/BestPractices/Multisite.php]: but no multisite `directories are` present

File: Check/BestPractices/Multisite.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  public function getAction() {
+    if ($this->score == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_FAIL) {
+      return dt('See https://pantheon.io/blog/drupal-multisite-much-ado-about-drupal-multisite for details.');
+    }
+    if ($this->score == SiteAuditCheckAbstract::AUDIT_CHECK_SCORE_WARN) {
+      if ($this->registry['multisite_enabled']) {
+        return dt('See https://www.drupal.org/node/2297419 for details on how to use multisite feature in Drupal 8.');
+      }
+      else {
+        return dt('Create sites.php file inside sites directory by copying example.sites.php file. See https://www.drupal.org/node/2297419 for more details.');
```
Comment (fluxsauce) [inline Check/BestPractices/Multisite.php]: `Inside the sites/ directory, copy example.sites.php to sites.php to create the configuration. See https://www.drupal.org/node/2297419 for details.`

File: Check/BestPractices/Services.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultInfo().
+   */
+  public function getResultInfo() {}
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultPass().
+   */
+  public function getResultPass() {
+    return dt('service.yml exists and is not a symbolic link.');
```
Comment (fluxsauce) [inline Check/BestPractices/Services.php]: services

File: Check/BestPractices/Services.php [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+   * Implements \SiteAudit\Check\Abstract\getDescription().
+   */
+  public function getDescription() {
+    return dt('Check if the services file exists.');
+  }
+
+  /**
+   * Implements \SiteAudit\Check\Abstract\getResultFail().
+   */
+  public function getResultFail() {}
```
Comment (fluxsauce) [inline Check/BestPractices/Services.php]: Missing a message; `services.yml does not exist! Copy the default.service.yml to services.yml and see https://www.drupal.org/documentation/install/settings-file for details.`

**Comment by fluxsauce** [comment]
Agree with the recommendations; please implement the SitesDefault and SitesSuperfluous checks and update the list of known default and expected files accordingly.

**Comment by shivanshuag** [comment]
Should there be a check for `{modules,themes,libraries,profiles}`? If yes, what should I name it?

**Comment by fluxsauce** [comment]
Skip `{modules,themes,libraries,profiles}` for now.

**Comment by fluxsauce** [comment]
Good job! In the future, please remove the "needs work" label when you've made the requested changes and it's ready for review again. Thank you for all your hard work!

**Pull request merged**